### DISCUSSION TITLE
What is the best practice for Effect to be executed only once onAppear?

### DISCUSSION THREAD

**Discussion by asa08** [pointfreeco/swift-composable-architecture#1635] [Q&A]
The easiest way is to have State flagged, but it is not cool.

```swift

import ComposableArchitecture

enum TestCore {}

extension TestCore {
  struct State: Equatable {
    var haveSent: Bool = false
  }
}

extension TestCore {
  enum Action: Equatable {
    case onAppear
  }
}

extension TestCore {
  struct Environment {
    let analyticsLogRepository: AnalyticsLogRepository
  }
}

extension TestCore {
  static let reducer = AnyReducer<State, Action, Environment>.combine(
    .init { state, action, _ in

      // sendLog() is a function that returns AnyPublisher
      func sendOnAppearLog() -> Effect<Action, Never> {
        if state.haveSent { return .none }
        state.haveSent = true
        return environment.analyticsLogRepository
          .sendLog()
          .fireAndForget()
      }

      switch action {
        case .onAppear:
          return sendOnAppearLog()
      }
    }
  )
}
```

What if I want to run `environment.analyticsLogRepository.sendLog()` only once the first time?

Any help is appreciated, thanks in advance!

**Comment by corysullivan** [comment]
Another option is to create a View modifier that only fires on the first onAppear.

See https://www.swiftjectivec.com/swiftui-run-code-only-once-versus-onappear-or-task/?utm_campaign=iOS%2BDev%2BWeekly&utm_medium=web&utm_source=iOS%2BDev%2BWeekly%2BIssue%2B583

**Reply by asa08** [reply]
Thanks for the advice :)
Other than onAppear, what do you do if you want to execute only the first Effect from multiple calls to an Action?

**Reply by mbrandonw** [reply]
For that you can always track some extra state:

```swift
struct State {
  var hasSentTapAction = false
  …
}
```

And then in the reducer:

```swift
case .tap:
  guard !state.hasSentTapAction
  else { return .none }

  return .run { send in … }
}
```

Or, if you want something more implicit, you could look at what we do for [effect cancellation][cancellation] and try to make your own custom effect that skips an effect if there is already one in flight.

[cancellation]: https://github.com/pointfreeco/swift-composable-architecture/blob/ff0e4e7684e1259992c0c3c396093ab8d0f0929a/Sources/ComposableArchitecture/Effects/Cancellation.swift
